freqtrade/data/dataprovider.py

```diff
@@ -63,9 +79,100 @@ def _set_cached_df(
+        :param candle_type: Any of the enum CandleType (must match trading mode!)
+        """
+        pair_key = (pair, timeframe, candle_type)
+
+        # For multiple leaders, if the data already exists, we'd merge
+        self.__external_pairs_df[pair_key] = (dataframe, datetime.now(timezone.utc))
+        self._set_data_event(pair_key)
+
+    def get_external_df(
+        self,
+        pair: str,
+        timeframe: str,
+        candle_type: CandleType
+    ) -> DataFrame:
+        """
+        Get the pair data from the external sources. Will wait if the policy is
+        set to, and data is not available.
+
+        :param pair: pair to get the data for
+        :param timeframe: Timeframe to get data for
+        :param candle_type: Any of the enum CandleType (must match trading mode!)
+        """
+        pair_key = (pair, timeframe, candle_type)
+
+        if pair_key not in self.__external_pairs_df:
+            self._wait_on_data(pair_key)
+
+            if pair_key not in self.__external_pairs_df:
+                return (DataFrame(), datetime.fromtimestamp(0, tz=timezone.utc))
+
+        return self.__external_pairs_df[pair_key]
+
+    def _set_data_event(self, key: PairWithTimeframe):
+        """
+        Depending on the WaitDataPolicy, if an event exists for this PairWithTimeframe
+        then set the event to release main thread from waiting.
+
+        :param key: PairWithTimeframe
+        """
+        pair_event = self.__external_pairs_event.get(key)
+
+        if pair_event:
+            num_concat, event = pair_event
+            self.__external_pairs_event[key] = (num_concat + 1, event)
+
+            if self._wait_data_policy == WaitDataPolicy.one:
+                logger.debug("Setting Data as policy is One")
+                event.set()
+            elif self._wait_data_policy == WaitDataPolicy.all and num_concat == self._num_sources:
+                logger.debug("Setting Data as policy is all, and is complete")
+                event.set()
+
+                del self.__external_pairs_event[key]
+
+    def _wait_on_data(self, key: PairWithTimeframe):
+        """
+        Depending on the WaitDataPolicy, we will create and wait on an event until
+        set that determines the full amount of data is available
+
+        :param key: PairWithTimeframe
+        """
+        if self._wait_data_policy is not WaitDataPolicy.none:
+            pair, timeframe, candle_type = key
+
+            pair_event = Event()
+            self.__external_pairs_event[key] = (0, pair_event)
+
+            timeout = self._wait_data_timeout \
+                if self._wait_data_policy is not WaitDataPolicy.all else 0
+
+            timeout_str = f"for {timeout} seconds" if timeout > 0 else "indefinitely"
+            logger.debug(f"Waiting for external data on {pair} for {timeout_str}")
+
+            if timeout > 0:
+                pair_event.wait(timeout=timeout)
+            else:
+                pair_event.wait()
```

Comment:
This is only acceptable if you enforce that waiting is limited to a maximum of 5-10 seconds.

waiting here will "stop the world" - which means freqtrade will no longer be handling open trades.
this therefore must change or be removed - as it's unacceptable.

Reply:
I think having this configurable is desired, especially for my usecase (even if we do limit it to a max time).

Reply:
Unless we can figure out a way to make sure the consumer just never receives data on time without syncing when the bots start.

Reply:
Blocking is inacceptable. That's not optional.

While you can block in the strategy - then it's your fault if trades are no longer processed- the freqtrade internals will not block.

Reply:
:+1: